File: server/src/services/daily_recommend_service.ts
import { PrismaClient } from '@prisma/client';
import { BusinessError } from '@/middlewares/error_middleware';
import { QuoteDetail, CategorySimple } from '@/types';

const prisma = new PrismaClient();

interface DailyRecommendItem {
    id: number;
    recommendDate: string;
    theme?: string;
    quotes: QuoteDetail[];
}

async function getToday(): Promise<DailyRecommendItem> {
    const todayStr = formatDate(new Date());

    let record = await prisma.dailyRecommend.findUnique({
        where: { recommendDate: startOfDay(new Date()) },
        include: {
            dailyRecommendQuotes: {
                orderBy: { sort: 'asc' },
                include: {
                    quote: { include: { quoteCategories: { include: { category: true } } } },
                },
            },
        },
    });

    if (!record) {
        record = await generateTodayRecommend();
    }

    return toDailyRecommendItem(record);
}

async function getByDate(date: string): Promise<DailyRecommendItem | null> {
    const d = new Date(date);
    if (isNaN(d.getTime())) throw new BusinessError(400, '日期格式无效');

    const record = await prisma.dailyRecommend.findUnique({
        where: { recommendDate: startOfDay(d) },
        include: {
            dailyRecommendQuotes: {
                orderBy: { sort: 'asc' },
                include: {
                    quote: { include: { quoteCategories: { include: { category: true } } } },
                },
            },
        },
    });

    if (!record) return null;
    return toDailyRecommendItem(record);
}

async function generateTodayRecommend() {
    // 跨数据库兼容：用 findMany 获取全部 id，JS 随机抽取（避免 RANDOM() 方言差异）
    const allQuotes = await prisma.quote.findMany({
        where: { isActive: true, auditStatus: 1 },
        select: { id: true },
    });
    if (allQuotes.length === 0) {
        throw new BusinessError(500, '暂无可推荐的名句');
    }

    // Fisher-Yates 随机抽取 10 个
    const pool = allQuotes.map((q) => q.id);
    const takeCount = Math.min(10, pool.length);
    const quoteIds: number[] = [];
    for (let i = 0; i < takeCount; i++) {
        const idx = Math.floor(Math.random() * pool.length);
        quoteIds.push(pool.splice(idx, 1)[0]);
    }

    const today = startOfDay(new Date());
    const record = await prisma.dailyRecommend.create({
        data: {
            recommendDate: today,
            theme: '今日精选',
            dailyRecommendQuotes: {
                create: quoteIds.map((qid, idx) => ({ quoteId: qid, sort: idx })),
            },
        },
        include: {
            dailyRecommendQuotes: {
                orderBy: { sort: 'asc' },
                include: {
                    quote: { include: { quoteCategories: { include: { category: true } } } },
                },
            },
        },
    });

    return record;
}

function toDailyRecommendItem(record: {
    id: number;
    recommendDate: Date;
    theme?: string | null;
    dailyRecommendQuotes: Array<{
        quote: {
            id: number;
            content: string;
            author?: string | null;
            source?: string | null;
            summary?: string | null;
            isFree: boolean;
            wordCount: number;
            pinyinData?: string | null;
            quoteCategories: Array<{ category: { id: number; name: string; type: string } }>;
        };
    }>;
}): DailyRecommendItem {
    const quotes: QuoteDetail[] = record.dailyRecommendQuotes.map((rq) => {
        const q = rq.quote;
        const categories: CategorySimple[] = q.quoteCategories.map((qc) => ({
            id: String(qc.category.id),
            name: qc.category.name,
            type: qc.category.type,
        }));
        return {
            id: String(q.id),
            content: q.content,
            author: q.author ?? undefined,
            source: q.source ?? undefined,
            summary: q.summary ?? undefined,
            isFree: q.isFree,
            categories,
            wordCount: q.wordCount,
            isLongText: q.wordCount > 100,
            pinyinData: q.pinyinData ?? undefined,
            collected: false,
            inRecitePlan: false,
        };
    });

    return {
        id: record.id,
        recommendDate: formatDate(record.recommendDate),
        theme: record.theme ?? undefined,
        quotes,
    };
}

function startOfDay(d: Date): Date {
    const x = new Date(d);
    x.setHours(0, 0, 0, 0);
    return x;
}

function formatDate(d: Date): string {
    const y = d.getFullYear();
    const m = String(d.getMonth() + 1).padStart(2, '0');
    const day = String(d.getDate()).padStart(2, '0');
    return `${y}-${m}-${day}`;
}

export { getToday, getByDate };

File: server-py/app/services/daily_recommend_service.py
"""每日推荐服务

对应 TypeScript 版 server/src/services/daily_recommend_service.ts
"""
import random as _random
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from ..deps import BusinessError
from ..models import DailyRecommend, DailyRecommendQuote, Quote, QuoteCategory, Category


def _start_of_day(d: datetime) -> datetime:
    return d.replace(hour=0, minute=0, second=0, microsecond=0)


def _format_date(d: datetime) -> str:
    return f"{d.year}-{d.month:02d}-{d.day:02d}"


def _get_quote_categories(db: Session, quote_id: int) -> list:
    qcs = db.query(QuoteCategory).filter(QuoteCategory.quoteId == quote_id).all()
    if not qcs:
        return []
    cat_ids = [qc.categoryId for qc in qcs]
    cats = db.query(Category).filter(Category.id.in_(cat_ids)).all()
    return [{"id": str(c.id), "name": c.name, "type": c.type} for c in cats]


def _to_daily_recommend_item(db: Session, record: DailyRecommend, rq_records: list) -> dict:
    quotes: list = []
    for rq in rq_records:
        q = db.query(Quote).filter(Quote.id == rq.quoteId).first()
        if not q:
            continue
        categories = _get_quote_categories(db, q.id)
        quotes.append({
            "id": str(q.id),
            "content": q.content,
            "author": q.author or None,
            "source": q.source or None,
            "summary": q.summary or None,
            "isFree": q.isFree,
            "categories": categories,
            "wordCount": q.wordCount,
            "isLongText": q.wordCount > 100,
            "pinyinData": q.pinyinData or None,
            "collected": False,
            "inRecitePlan": False,
        })

    return {
        "id": record.id,
        "recommendDate": _format_date(record.recommendDate),
        "theme": record.theme or None,
        "quotes": quotes,
    }


def _load_with_quotes(db: Session, record: DailyRecommend) -> list:
    return (
        db.query(DailyRecommendQuote)
        .filter(DailyRecommendQuote.recommendId == record.id)
        .order_by(DailyRecommendQuote.sort.asc())
        .all()
    )


def getToday(db: Session) -> dict:
    today = _start_of_day(datetime.now())

    record = db.query(DailyRecommend).filter(DailyRecommend.recommendDate == today).first()
    if not record:
        record = _generate_today_recommend(db, today)

    rq_records = _load_with_quotes(db, record)
    return _to_daily_recommend_item(db, record, rq_records)


def getByDate(db: Session, date: str) -> dict | None:
    try:
        d = datetime.fromisoformat(date)
        if d.hour == 0 and d.minute == 0 and d.second == 0:
            pass  # already a date-only
    except (ValueError, TypeError):
        try:
            d = datetime.fromisoformat(date + "T00:00:00")
        except (ValueError, TypeError):
            raise BusinessError(400, "日期格式无效")

    d = _start_of_day(d)
    record = db.query(DailyRecommend).filter(DailyRecommend.recommendDate == d).first()
    if not record:
        return None

    rq_records = _load_with_quotes(db, record)
    return _to_daily_recommend_item(db, record, rq_records)


def _generate_today_recommend(db: Session, today: datetime) -> DailyRecommend:
    all_quotes = db.query(Quote).filter(
        Quote.isActive == True,
        Quote.auditStatus == 1,
    ).with_entities(Quote.id).all()

    if not all_quotes:
        raise BusinessError(500, "暂无可推荐的名句", 500)

    pool = [q[0] for q in all_quotes]
    take_count = min(10, len(pool))
    quote_ids: list[int] = []
    for _ in range(take_count):
        idx = _random.randint(0, len(pool) - 1)
        quote_ids.append(pool.pop(idx))

    record = DailyRecommend(
        recommendDate=today,
        theme="今日精选",
        status=1,
    )
    db.add(record)
    db.commit()
    db.refresh(record)

    for idx, qid in enumerate(quote_ids):
        rq = DailyRecommendQuote(
            recommendId=record.id,
            quoteId=qid,
            sort=idx,
        )
        db.add(rq)
    db.commit()

    return record
